### ISSUE TITLE
More Validation Rules Support

### ISSUE THREAD

**Issue by bethanyamber** [knuckleswtf/scribe#140]
Currently when using Laravel FormRequests, validation rules can be extracted and used in the Scribe generated docs, which is really useful. However as the docs mention, there are quite a lot of validation rules not currently supported  by Scribe, some of which would really be useful.

These are rules such as:

- min/max

- before/before or equal (date)

- after/after or equal (date)

- required if/unless/with/without (also mentioned in #111 )

- exists

If support for these validation rules could be included that would be really helpful to what I am currently working on!

**Comment by shalvah** [comment]
The reason most of these rules are not supported of because they're non-deterministic, and with the way Scribe works now, it would need a lot of backtracking to process them correctly.

Note that Scribe uses these rules for two things:
- Generating a description for a parameter
- Generating a valid example of a parameter

## Polymorphic rules (min/max)
Let's take the min rule as an example. It can mean several things:
- minimum length (for a string)
- minimum size (for a file)
- minimum value (for a number)

Assuming we have parameter `thing`, with that rule set. When Laravel is validating that rule, it doesn't have to worry about that, because it's either `thing` is missing in the request (so it's skips the rule entirely), or `thing` is present, in which cas it can check the type of `thing` and know what interpretation of the rule to apply.

When Scribe is parsing this rule, it doesn't have a valid value to use. Maybe you provided one as an example, but that isn't always guaranteed. So, either:
- Scribe guesses the type of `thing`, or
- Scribe looks for another rule in your list of rules for `thing` that specifies the type. This now brings up another problem. Scribe has to pause or end the current parsing run and switch to that one, before returning to this. See the difficulty?

## Dependent rules
The `before` rule can take a date or the name of another field. The date part is straightforward. But when before is set to another field, what happens? The description part is simple. But when Scribe is trying to generate a valid value for `thing`, there's an added complication: it must ensure that the value it generates comes before the value for `thing2`.

But what if `thing2` hasn't been processed yet? Scribe has to leap out of the current processing point and generate a value for `thing2` first, before returning.

There's a similar problem with `required_if` and friends. When Scribe is generating examples, it complicates the whole process, because encountering a dependent parameter means going back or forward to check on the independent ones first.

## Database rules
`exists` depends on your database connection to b get a valid value. This probably only matters for response calls, but I guess that can be worked around.

# Long story short...
All of this is doable, but it will take a lot of careful planning, restructuring and work, otherwise you'll end up with [bugs all over the place](https://github.com/mpociot/laravel-apidoc-generator/issues/343#issue-367508887). If I have the time, I may explore this for v3.

One thing we could do for a start is parse some rules only for the description, while still generating possibly invalid example values. I didn't want to do this, because I wanted to avoid confusing readers ("the description says this field must be after this date, but in this example, they aren't?!")

**Comment by matyhaty** [comment]
Thanks  @shalvah 

There are a number of validation rules which are not 'dependency' validation rules (e.g. Min) which would be good to add in. (Hopefully far more simple)

Would you be able to document the process of writing new ones briefly and I will look at getting more of them written in, and do a PR

Hopefully that will help us use Scribe and contribute to the project.

Any simple guide or point in the right direction would be excellent!

**Comment by bmbBAMBUS** [comment]
Maybe solution for `required_if` is to pass string to `field-details.blade.php` instead of bool, and pass desired string for this field in `bodyParameters()` function ?

```
return [
   'fieldName' => [
      'required' => 'optional|required_if:anotherField,not_present',
      'description' => 'Description',
  ]
]
```

**Comment by shalvah** [comment]
I think we can do it, but it will need some rethinking of the current approach. With the right data structure and logic, we might be able to handle field-dependent rules in v3.

**Comment by shalvah** [comment]
Update: Polymorphic rules (min, max, between, size) are now working in v3. Dependent rules (required_*, before, after) have partial support (ie they are used for generating descriptions, but not necessary for the example.)
